enhancements/machine-api/converting-machine-api-to-cluster-api.md

```diff
@@ -0,0 +1,825 @@
+
+In this solution, the Machine API controllers would be rewritten to convert Machine API resources internally to Cluster API resources.
+Once the conversion has been completed, the controllers would feed the converted Cluster API resources into the Cluster API methods, using them as a library.
+
+This alternative has been dismissed based on:
+- No migration from Machine API to Cluster API
+- No reduction in maintenance burden
+- No longer term plan to remove the Machine API controllers
+- It is unclear how to handle conversion when there is a feature gap in Cluster API
+
+#### Convert resources using a webhook
+
+Rather than using a controller to convert the resources after they have been accepted by the API server, we could use a ValidatingAdmissionWebhook to convert resources as updates are applied.
+This has the benefit of applying updates synchronously to both API groups.
+
+However, it also adds additional complexity"
+- If the conversion webhook accepts the change and writes out the Cluster API resources
+  - Does this then call the converse webhook and start to write out the Machine API resources, creating a loop?
+  - What happens if a later admission webhook rejects the update?
+  - What end user webhooks are built on top of our extensions, may this interfere with them?
+
+#### Offline conversion
+
+Rather than converting resources within the cluster, we could provide users with a tool to offline convert their Machine API resources to Cluster API.
+This is possible even with the current proposal as an additional feature which may be desirable for some users to create additional MachineSets, based on existing MachineSets.
+
+However, this doesn't help with the goal of removing Machine API controllers in the longer term, so is not sufficient to meet our goals on its own.
+
+### Filtering events in Cluster API
+
+#### Filtering reconciles in Cluster API controllers
+
+To enable the Cluster API controllers to observe the authoritative API, we would need to implement a check at the beginning of the reconcile logic of each controller.
+
+This check would look up the current state of the Machine API equivalent resource and determine whether the controller should be reconciling that specific resource or not.
+
+To ensure we always have the latest version, caches should not be used when performing this check (Cluster API is based on controller-runtime which by default uses a caching client).
+
+Each of the infrastructure Machine controllers will not have an equivalent resource to look up, and therefore must first identify the Machine that they are associated with and look up the equivalent resource for this Machine.
+
+This would require carrying a patch in the Cluster API controllers to implement this logic and would create toil for maintenance.
+
+##### Filtering reconciles via a webhook
+
+An alternative to the above is to have each Cluster API controller perform an equivalent of a SubjectAccessReview.
```

Comment:
Just to be clear, is this the approach you're suggesting?

Reply:
No, alternatives are alternatives. The actual approach should be described in the body of the proposal. In this case, the pausing and unpausing of resources across the two API groups is the actual proposal

Reply:
Ack. Reading as markdown has got me here :P